### DISCUSSION TITLE
concise and reusuable non-empty array check after transform

### DISCUSSION THREAD

**Discussion by ThomasFen** [colinhacks/zod#3191] [Q&A] [answered]
My goal is that the validated TestData should be of type `[ string, string[] ]`, without doing the validation within transform callback.
I came up with 2 solutions (see code), but I'm not happy with either.

1. Doing the validation with Zod's `.nonEmpty` will validate the elements (and if it's an array) a second time.
2. Using a generic function is verbose

What's the proper way to achieve what I want?

```ts
const testData = ['1', '2', '3'];

/**
 * Without any non-empty validation
 */
const schema = z
  .array(z.string())
  .transform((elements) => elements.filter((el) => el != '3'));
const validatedTestData = schema.parse(testData);
// type validatedTestData = string[]

/**
 * This works but does redundant array item validation
 */
const schemaRedundant = schema.pipe(z.array(z.string()).nonempty());
const validatedTestDataRedundant = schemaRedundant.parse(testData);
// type validatedTestData = [string, ...string[]]

/**
 * This also works but is verbose
 */
type OutputType<T> = T extends z.ZodEffects<z.ZodTypeAny, infer O> ? O : never;
type Flatten<T> = T extends Array<infer Item> ? Item : never;

const requireElements = <T extends z.ZodEffects<z.ZodTypeAny, unknown[]>>(
  schema: T
) =>
  schema.refine(
    (arr): arr is [Flatten<OutputType<T>>, Flatten<OutputType<T>>[]] =>
      arr.length > 0
  );
const validatedTestDataWithGenericFunction =
  requireElements(schema).parse(testData);
// type validatedTestDataWithGenericFunction = [string, ...string[]]

```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const schemaBase = z.string().array().nonempty()
const schema = schemaBase
    .transform( ( elements ) => elements.filter( ( el ) => el != '3' ) )
    .pipe( schemaBase )
const validatedTestData = schema.parse( [ '1', '2', '3' ] )
// const validatedTestData: [string, ...string[]]

console.log( validatedTestData ) // [ '1', '2' ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by ThomasFen** [reply]
thanks for your contribution, but that's similar to my first solution. The `transform` doesn't change the input type, so I want to avoid the unnecessary second `z.string().array()` check

**Reply by JacobWeisenburger** [reply]
why are you concerned about making a 2nd `z.string().array()` check?

**Reply by ThomasFen** [reply]
If the array contains elements other than strings, necessitating intricate validation logic, re-executing such logic may incur considerable computational costs

**Reply by JacobWeisenburger** [reply]
https://www.youtube.com/watch?v=tKbV6BpH-C8
